### PR TITLE
Add support kind version also adds significant coverage to deploy

### PR DESCRIPTION
**Pull request by marcushines**

### CODE REVIEW

File: deploy/deploy.go [hunk 1]
```diff
@@ -222,6 +223,56 @@ type KindSpec struct {
 	AdditionalManifests      []string          `yaml:"additionalManifests"`
 }
 
+var (
+	vOut                  = bytes.NewBuffer([]byte{})
+	vExec execerInterface = kexec.NewExecer(vOut, vOut)
+)
+
+type version struct {
+	Major int
```
Comment (alexmasi) [inline deploy/deploy.go:232]: Do these need to be exported?

File: deploy/deploy.go [hunk 2]
```diff
@@ -233,7 +284,36 @@ func (k *KindSpec) checkDependencies() error {
 			errs.Add(fmt.Errorf("install dependency %q to deploy", bin))
 		}
 	}
-	return errs.Err()
+	if errs.Err() != nil {
+		return errs.Err()
+	}
+	if k.Version != "" {
```
Comment (alexmasi) [inline deploy/deploy.go:290]: Where does Version even come into play? Currently its a field in the yaml but I can't find where it is used. 

If the user does not provide a version what should happen?

File: deploy/deploy.go [hunk 3]
```diff
@@ -233,7 +284,36 @@ func (k *KindSpec) checkDependencies() error {
...
+		}
+
+		gotV, err := k.getKindVersion(vKindFields[1])
+		if err != nil {
+			return fmt.Errorf("kind version check failed: %w", err)
+		}
+		if gotV.Less(wantV) {
+			return fmt.Errorf("kind version check failed: got %s, want %s", gotV, wantV)
+		}
+		log.Infof("kind version valid: got %s want %s", gotV, wantV)
```
Comment (alexmasi) [inline deploy/deploy.go:314]: above (line 112) we check the existence of docker and kubectl, we should also check those versions (and likely need to check against the Image field)

File: deploy/deploy.go [hunk 4]
```diff
@@ -233,7 +284,36 @@ func (k *KindSpec) checkDependencies() error {
...
+		if len(vKindFields) < 2 {
+			return fmt.Errorf("failed to parse kind version from: %s", vOut)
+		}
+
+		gotV, err := k.getKindVersion(vKindFields[1])
+		if err != nil {
+			return fmt.Errorf("kind version check failed: %w", err)
+		}
+		if gotV.Less(wantV) {
+			return fmt.Errorf("kind version check failed: got %s, want %s", gotV, wantV)
```
Comment (alexmasi) [inline deploy/deploy.go:312]: so the only binary that we would upgrade automatically would be kubectl in the Kind case?

File: deploy/deploy.go (hunk 2 above)
Comment (marcushines) [reply deploy/deploy.go:290]: it is used right below?

Comment (marcushines) [reply deploy/deploy.go:290]: if none set - do nothing assume the user knows what they are doing

File: deploy/deploy.go (hunk 3 above)
Comment (marcushines) [reply deploy/deploy.go:314]: this is just step one

File: deploy/deploy.go (hunk 4 above)
Comment (marcushines) [reply deploy/deploy.go:312]: others will come later 
each should have an option to leave them or to upgrade them if they are older

File: deploy/deploy.go (hunk 1 above)
Comment (marcushines) [reply deploy/deploy.go:232]: i am just signally to users that those 3 fields are usable

**Review by alexmasi — APPROVED** [review]

**Pull request merged**